android/modules/ui/src/java/ti/modules/titanium/ui/ScrollableViewProxy.java

```diff
@@ -66,164 +65,153 @@ public ScrollableViewProxy()
-					handled = true;
-				}
-				break;
 			default:
 				handled = super.handleMessage(msg);
 		}
 
 		return handled;
 	}
 
-	@Kroll.getProperty
-	public Object getViews()
+	public void removeAllViews()
 	{
-		TiViewProxy[] childViewArray = new TiViewProxy[0];
-		TiUIScrollableView view = getView();
-		if (view != null) {
-			childViewArray = view.getViews().toArray(childViewArray);
+		for (final TiViewProxy view : this.views) {
+			view.releaseViews();
+			view.setParent(null);
+		}
+		this.views.clear();
+
+		if (scrollableView != null) {
+			scrollableView.getAdapter().notifyDataSetChanged();
 		}
-		return childViewArray;
+	}
+
+	public ArrayList<TiViewProxy> getViewsList()
+	{
+		return (ArrayList<TiViewProxy>) this.views;
+	}
+
+	@Kroll.getProperty
+	public TiViewProxy[] getViews()
+	{
+		return this.views.toArray(new TiViewProxy[0]);
 	}
 
 	@Kroll.setProperty
-	public void setViews(Object viewsObject)
+	public void setViews(Object views)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_SET_VIEWS), viewsObject);
+		removeAllViews();
+
+		if (views instanceof Object[]) {
+			for (final Object view : (Object[]) views) {
+				if (view instanceof TiViewProxy) {
+					addView((TiViewProxy) view);
+				}
+			}
+		}
 	}
 
 	@Kroll.method
-	public void addView(Object viewObject)
+	public void addView(TiViewProxy view)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_ADD_VIEW), viewObject);
+		if (!this.views.contains(view)) {
+			view.setActivity(getActivity());
+			view.setParent(this);
+			this.views.add(view);
+
+			if (scrollableView != null) {
+				scrollableView.getAdapter().notifyDataSetChanged();
+			}
+		}
 	}
 
 	@Kroll.method
 	public void insertViewsAt(int insertIndex, Object viewObject)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_INSERT_VIEWS_AT, insertIndex, 0),
-											viewObject);
+		if (viewObject instanceof TiViewProxy) {
+			final TiViewProxy view = (TiViewProxy) viewObject;
+
+			if (!this.views.contains(view)) {
+				view.setActivity(getActivity());
+				view.setParent(this);
+				this.views.add(insertIndex, view);
+			}
+
+		} else if (viewObject instanceof Object[]) {
+			Object[] views = (Object[]) viewObject;
+
+			for (int i = 0; i < views.length; i++) {
+				insertViewsAt(insertIndex, views[i]);
+			}
+		} else {
+
+			// Skip updating adapter.
+			return;
+		}
+
+		if (scrollableView != null) {
+			scrollableView.getAdapter().notifyDataSetChanged();
+		}
 	}
 
 	@Kroll.method
 	public void removeView(Object viewObject)
 	{
-		TiMessenger.sendBlockingMainMessage(getMainHandler().obtainMessage(MSG_REMOVE_VIEW), viewObject);
+		if (viewObject instanceof Number) {
+			this.views.remove((int) viewObject);
+
+		} else if (viewObject instanceof TiViewProxy) {
+			final TiViewProxy view = (TiViewProxy) viewObject;
+
+			this.views.remove(view);
+
+		} else {
+
+			// Skip updating adapter.
+			return;
+		}
```

Comment:
There's a bug here where if property "currentPage" references the last page and you remove via this method, the "currentPage" property does not get updated and becomes invalid. This can cause an exception to be thrown if you use this invalid index.

You can reproduce this with my `ScrollableViewAddRemoveTest.js` script attached to [TIMOB-28459](https://jira.appcelerator.org/browse/TIMOB-28459).
1. Scroll to the last page.
2. Tap on the Remove button.
3. Select dialog's "Current" option to delete current page. ("currentPage" is now invalid.)
4. Tap on the Remove button again.
5. Select dialog's "Current" option again.
6. Notice an exception is thrown due to an invalid index.